elftools/ehabi/decoder.py

```diff
@@ -0,0 +1,284 @@
+        #  uint8_t Start = ((Opcode1 & 0xf0) >> 4);
+        #  uint8_t Count = ((Opcode1 & 0x0f) >> 0);
+        #  PrintRegisters((((1 << (Count + 1)) - 1) << Start), "wR");
+        self._index += 1
+        op1 = self._bytecode_array[self._index]
+        self._index += 1
+        start = ((op1 & 0xf0) >> 4)
+        count = ((op1 & 0x0f) >> 0)
+        return 'pop %s' % self._print_registers((((1 << (count + 1)) - 1) << start), "wR")
+
+    def _decode_11000111_0000iiii(self):
+        #   SW.startLine()
+        #     << format("0x%02X 0x%02X ; %s", Opcode0, Opcode1,
+        #               ((Opcode1 & 0xf0) || Opcode1 == 0x00) ? "spare" : "pop ");
+        #   if ((Opcode1 & 0xf0) == 0x00 && Opcode1)
+        #       PrintRegisters(Opcode1 & 0x0f, "wCGR");
+        self._index += 1
+        op1 = self._bytecode_array[self._index]
+        self._index += 1
+        if (op1 & 0xf0) != 0 or op1 == 0x00:
+            return 'spare'
+        else:
+            return 'pop %s' % self._print_registers(op1 & 0x0f, "wCGR")
+
+    def _decode_11001000_sssscccc(self):
+        #   SW.startLine() << format("0x%02X 0x%02X ; pop ", Opcode0, Opcode1);
+        #   uint8_t Start = 16 + ((Opcode1 & 0xf0) >> 4);
+        #   uint8_t Count = ((Opcode1 & 0x0f) >> 0);
+        #   PrintRegisters((((1 << (Count + 1)) - 1) << Start), "d");
+        self._index += 1
+        op1 = self._bytecode_array[self._index]
+        self._index += 1
+        start = 16 + ((op1 & 0xf0) >> 4)
+        count = ((op1 & 0x0f) >> 0)
+        return 'pop %s' % self._print_registers((((1 << (count + 1)) - 1) << start), "d")
+
+    def _decode_11001001_sssscccc(self):
+        #   SW.startLine() << format("0x%02X 0x%02X ; pop ", Opcode0, Opcode1);
+        #   uint8_t Start = ((Opcode1 & 0xf0) >> 4);
+        #   uint8_t Count = ((Opcode1 & 0x0f) >> 0);
+        #   PrintRegisters((((1 << (Count + 1)) - 1) << Start), "d");
+        self._index += 1
+        op1 = self._bytecode_array[self._index]
+        self._index += 1
+        start = ((op1 & 0xf0) >> 4)
+        count = ((op1 & 0x0f) >> 0)
+        return 'pop %s' % self._print_registers((((1 << (count + 1)) - 1) << start), "d")
+
+    def _decode_11001yyy(self):
+        self._index += 1
+        return 'spare'
+
+    def _decode_11000nnn(self):
+        #   SW.startLine() << format("0x%02X      ; pop ", Opcode);
+        #   PrintRegisters((((1 << ((Opcode & 0x07) + 1)) - 1) << 10), "wR");
+        opcode = self._bytecode_array[self._index]
+        self._index += 1
+        return 'pop %s' % self._print_registers((((1 << ((opcode & 0x07) + 1)) - 1) << 10), "wR")
+
+    def _decode_11010nnn(self):
+        #   SW.startLine() << format("0x%02X      ; pop ", Opcode);
+        #   PrintRegisters((((1 << ((Opcode & 0x07) + 1)) - 1) << 8), "d");
+        opcode = self._bytecode_array[self._index]
+        self._index += 1
+        return 'pop %s' % self._print_registers((((1 << ((opcode & 0x07) + 1)) - 1) << 8), "d")
```

Comment:
I see this subexpression a lot:
`self._print_registers((((1 << ((X) + 1)) - 1) << Y)`
Do you think creating a helper method that accepts parameters `X`, and `Y` would help [DRY up](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) this code?

Reply:
I reviewed every cross reference of `_print_registers` and `_print_GPR`. Most of the input parameter are different. Only 3 of them are the same. I think it's unnecessary to create sub_method for them.

BTW, I would like to follow the computing from `llvm/tools/llvm-readobj`.